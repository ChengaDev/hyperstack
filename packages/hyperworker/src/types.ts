import type {
  MailerSettings,
  WorkerBackendOpts,
  WorkersMode,
} from '@hyperstackjs/typings'
import type { HyperWorker } from './worker'
import type { Mailer } from './mailer'
export { MailerSettings, WorkerBackendOpts, WorkersMode }
export interface Worker {
  perform(args: any): any
}
export interface WorkerTypes {
  HyperWorker: typeof HyperWorker
  Mailer: typeof Mailer
}
export interface LoadWorkersOpts {
  workersGlob: string
  mailersGlob: string
}

export interface CreateWorkerConnectionOpts {
  redisUrl: string
  rejectUnauthorized?: boolean
}

export interface BuildWorkersParams {
  isInProcess: boolean
  workersGlob: string
  mailersGlob: string
  mailerTemplatesRoot: string
  mailerSettings?: MailerSettings
  logger: any
  redisUrl?: string
  backendOpts?: WorkerBackendOpts
  truncate?: boolean
  beforeWorkers?: (_p: any) => void
  afterWorkers?: (_p: any) => void
}
