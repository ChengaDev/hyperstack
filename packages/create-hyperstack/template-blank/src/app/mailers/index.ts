// suggested mailer layout:
// mailer-name.ts
// mailer-name/
//  template-contents
//
// export your mailers here:
// export { MyMailer } from './my-mailer'

export {}
