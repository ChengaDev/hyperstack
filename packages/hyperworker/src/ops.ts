import type { HyperWorker } from './worker'

export const getJobCounts = async (workers: typeof HyperWorker[]) => {
  const counts = []
  for (const W of workers) {
    const wc = await W.jobCounts()
    counts.push({ name: W.name, queue: W.queueName, ...wc })
  }
  return counts
}
