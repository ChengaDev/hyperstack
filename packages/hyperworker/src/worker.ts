import 'reflect-metadata'
import createDebug from 'debug'
import type { Logger } from '@hyperstackjs/typings'
import type { BullmqBackend, JobCounts } from './backend'
import type { Worker } from './types'
const debug = createDebug('hyperworker:worker')

const K_QUEUE_NAME = 'hyperworker:queue-name'
const K_PRODUCER_ONLY = 'hyperworker:queue-producer-only'

export const queueAs = (queueName: string) => (target: Object) => {
  return Reflect.defineMetadata(K_QUEUE_NAME, queueName, target)
}

export const producerOnly = () => (target: Object) => {
  return Reflect.defineMetadata(K_PRODUCER_ONLY, 'true', target)
}

class HyperWorker implements Worker {
  static inprocess = false

  static logger: Logger | null = null

  static backend: BullmqBackend | null = null

  static enqueue: any = null

  static queue: any = null

  static reportError = (_err: Error, _tags?: any) => {}

  static clear() {
    this.logger = null
    this.backend = null
    this.inprocess = false
    this.enqueue = null
    this.reportError = (_err, _tags?) => {}
  }

  static initWorker({ logger }: any) {
    this.logger = logger
  }

  static setMode({
    inprocess,
    reportError,
  }: {
    inprocess: boolean
    reportError?: (err: Error, tags?: any) => void
  }) {
    this.inprocess = inprocess
    if (reportError) {
      this.reportError = reportError
    }
  }

  static setTest(reportError: any) {
    this.setMode({ inprocess: true, reportError })
  }

  static get queueName() {
    return Reflect.getOwnMetadata(K_QUEUE_NAME, this) || this.name.toLowerCase()
  }

  static get producerOnly() {
    return Reflect.getOwnMetadata(K_PRODUCER_ONLY, this) || false
  }

  static async performNow(args: any) {
    const w = new this()
    return w.perform({ data: args })
  }

  static async performLater(_args: any) {
    if (this.inprocess) {
      debug(`${this.name}: perform now '%o'`, _args)
      return this.performNow(_args)
    }
    debug(`${this.name}: perform later '%o'`, _args)
    return this.enqueue(_args, {}) // TODO opts
  }

  static async jobCounts() {
    return (await this.queue.getJobCounts()) as JobCounts
  }

  get logger() {
    return HyperWorker.logger
  }

  perform(_args: any): any {
    // tbd
  }
}

export { HyperWorker }
