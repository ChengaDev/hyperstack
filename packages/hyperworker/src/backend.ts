import L from 'lodash'
import { Worker as BMWorker, Queue } from 'bullmq'
import createDebug from 'debug'
import type { WorkerBackendOpts } from '@hyperstackjs/typings'
import { HyperWorker } from './worker'
export interface JobCounts {
  waiting: number
  active: number
  completed: number
  failed: number
  delayed: number
}

const debug = createDebug('hyperworker:backend')

const sleep = (t: number) =>
  new Promise((resolve, _reject) => setTimeout(() => resolve(true), t))

const defaultWorkerOpts = {
  lockDuration: 30000,
}

const defaultQueueOpts = {
  // make data sizes predictable, modest, and auto healing
  streams: {
    events: { maxLen: 1000 }, // default was 10k
  },
  defaultJobOptions: {
    removeOnComplete: true, // default never
    removeOnFail: { count: 100 }, // default never
    stackTraceLimit: 5, // default no limit
  },
}
class BullmqBackend {
  // our worker types, for referece
  private workers?: any[]

  // bullmq workers, will 'new' our worker types and give them work
  // they won't work unless calling 'start'
  private bmworkers?: any[]

  // bullmq queues, gives worker types an 'enqueue' method
  private bmqueues: any[]

  private opts: WorkerBackendOpts

  public connection: any

  constructor(workers: any[], connection: any, opts?: WorkerBackendOpts) {
    this.connection = connection
    this.workers = workers
    this.opts = {
      workerOpts: L.merge({}, defaultWorkerOpts, opts?.workerOpts),
      queueOpts: L.merge({}, defaultQueueOpts, opts?.queueOpts),
    }

    debug('opts: %o', this.opts)
    this.bmqueues = workers.map((Klass) => {
      debug(`adding queue: ${Klass.queueName} from class ${Klass.name}`)
      const q = new Queue(Klass.queueName, {
        connection: this.connection,
        ...this.opts.queueOpts,
      })
      q.on('error', (err) => {
        HyperWorker.logger && HyperWorker.logger.error(`[error] ${err}`)
        HyperWorker.reportError(err)
      })
      const enqueue = (data: any, opts: any) => q.add('job', data, opts)
      Klass.enqueue = enqueue
      Klass.queue = q
      return q
    })
    this.workers = workers
  }

  async processJobs() {
    if (this.bmworkers) {
      throw new Error('already processing jobs')
    }
    if (!HyperWorker.logger) {
      throw new Error('logger is missing. did the app boot correctly?')
    }
    const { magenta, bold } = HyperWorker.logger.colors
    const pref = bold(magenta('workers:'))
    if (!this.workers) {
      throw new Error(
        "you need to wire into your worker first. call 'wireInto(<workers>)'"
      )
    }
    this.bmworkers = this.workers
      .filter((Klass) => {
        if (Klass.producerOnly) {
          HyperWorker.logger!.info(
            `${pref} producer only job: ${Klass.queueName}`
          )
          return false
        }
        return true
      })
      .map((Klass) => {
        debug(`processing queue: ${Klass.queueName} from class ${Klass.name}`)
        return new BMWorker(
          Klass.queueName,
          async (job) => {
            debug(`processing job in: ${Klass.queueName}`)
            const w = new Klass()
            await w.perform(job)
          },
          {
            connection: this.connection,
            ...this.opts.workerOpts,
          }
        )
      })
    this.bmworkers.forEach((w) => {
      w.on('drained', (_job: any) => {
        HyperWorker.logger!.info(`${pref} drained: ${w.name}`)
      })

      w.on('completed', (_job: any) => {
        HyperWorker.logger!.info(`${pref} completed: ${w.name}`)
      })

      w.on('failed', (job: any, err: Error) => {
        HyperWorker.logger!.info(`${pref} failed: ${w.name} (${err})`)
        HyperWorker.reportError(err, { worker: w.name, jobId: job.id })
      })
      w.on('error', (err: Error) => {
        HyperWorker.logger!.error(`${pref} error: ${w.name} (${err})`)
        HyperWorker.reportError(err, { worker: w.name })
      })
    })
  }

  async stop() {
    if (this.bmworkers) {
      for (const w of this.bmworkers) {
        if (w) {
          w.connection.removeAllListeners('close')
          w.connection.removeAllListeners('error')
          w.removeAllListeners('error')
          w.removeAllListeners('completed')
          w.removeAllListeners('drained')
          await w.close()
        }
      }
    }
    if (this.bmqueues) {
      for (const q of this.bmqueues) {
        if (q) {
          q.connection.removeAllListeners('close')
          q.connection.removeAllListeners('error')
          q.removeAllListeners('error')
          q.removeAllListeners('completed')
          q.removeAllListeners('drained')
          await q.close()
        }
      }
    }
    await this.connection.quit()
    this.workers = undefined

    await sleep(500) // give some time for emitters to drain
  }
}

export { BullmqBackend }
