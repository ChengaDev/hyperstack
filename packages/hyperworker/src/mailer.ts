/* eslint-disable @typescript-eslint/no-var-requires */
import nodemailer from 'nodemailer'
import createDebug from 'debug'
// @ts-expect-error 7016
import { interactsWithMail, stubTransport } from 'nodemailer-stub'
import { HyperWorker, queueAs } from './worker'
import type { MailerSettings } from './types'
const debug = createDebug('hyperworker:mailer')

/*
1. renderer (takes template path, then render takes params)
2. transport (nodemailer)
*/

class MailDelivery {
  constructor(private MailerType: typeof Mailer, private details: any) {}

  async deliver() {
    if (!HyperWorker.logger) {
      throw new Error('Logger is not configured. Did the app boot correctly?')
    }

    const { cyan, bold } = HyperWorker.logger.colors
    const pref = bold(cyan('mailers:'))
    HyperWorker.logger.info(
      `${pref} ${this.MailerType.name} delivering`,
      this.details
    )
    return this.MailerType.performNow(this.details)
  }

  async deliverLater() {
    if (!HyperWorker.logger) {
      throw new Error('Logger is not configured. Did the app boot correctly?')
    }
    const { cyan, bold } = HyperWorker.logger.colors
    const pref = bold(cyan('mailers:'))
    HyperWorker.logger.info(
      `${pref} ${this.MailerType.name} delivering later`,
      this.details
    )
    return this.MailerType.performLater(this.details)
  }
}

@queueAs('mails')
class Mailer extends HyperWorker {
  static transport: any | null = null

  static send = false

  static preview = false

  static templatesRoot: string | null = null

  // <testing> mailer stub only convenience
  static clearDeliveries() {
    interactsWithMail.flushMails()
  }

  static deliveries() {
    return interactsWithMail.mails
  }

  static lastMail() {
    return interactsWithMail.lastMail()
  }
  // </testing>

  static clear(): void {
    this.transport = null
    this.send = false
    this.preview = false
    this.templatesRoot = null
  }

  static defaults = {
    from: 'Joe <mailer@example.com>',
  }

  static initMailer({
    templatesRoot,
    settings,
  }: {
    templatesRoot: string
    settings: MailerSettings
  }) {
    this.templatesRoot = templatesRoot
    if (settings) {
      this.send = !!settings.send
      this.preview = !!settings.preview
      if (settings.delivery === 'test') {
        this.transport = stubTransport
      } else if (settings.delivery === 'smtp') {
        this.transport = nodemailer.createTransport(settings.smtpSettings)
      }
    }
  }

  static setNone() {
    this.send = false
    this.preview = false
    this.transport = null
  }

  static setTest() {
    this.send = true
    this.preview = false
    this.transport = stubTransport
  }

  static setDevelopment() {
    this.send = false
    this.preview = true
    this.transport = null
  }

  static setProduction() {
    this.send = true
    this.preview = false
  }

  async perform({ data }: any) {
    if (!HyperWorker.logger) {
      throw new Error('Logger is not configured. Did the app boot correctly?')
    }
    if (!Mailer.templatesRoot) {
      throw new Error('please set Mailer.templatesRoot')
    }
    // lazy because it takes lots of time to load eagerly
    const Email = require('email-templates')
    debug(`mailer: send settings '%o'`, {
      send: Mailer.send,
      preview: Mailer.preview,
    })
    const tmpl = new Email({
      views: { root: Mailer.templatesRoot, options: { extension: 'ejs' } },
      transport: Mailer.transport || { jsonTransport: true },
      send: Mailer.send,
      preview: Mailer.preview,
      juice: false,
    })

    try {
      debug(`mailer: perform '%o'`, data)
      const res = await tmpl.send(data)

      const { cyan, bold } = HyperWorker.logger.colors
      const pref = bold(cyan('mailers:'))
      HyperWorker.logger.info(
        `${pref} delivered email to: ${data.message.to}`,
        res
      )
      return res
    } catch (e: any) {
      HyperWorker.logger.error(e)
      HyperWorker.reportError(e)
    }
  }

  static mail(details: any, _opts: any = {}) {
    details.message = { ...this.defaults, ...details.message }
    return new MailDelivery(this, details)
  }

  get logger() {
    return HyperWorker.logger
  }
}

export { Mailer, MailDelivery }
