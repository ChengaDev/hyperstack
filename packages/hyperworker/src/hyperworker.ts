/* eslint-disable @typescript-eslint/no-var-requires */
import glob from 'glob'
import L from 'lodash'
import createDebug from 'debug'
import { HyperWorker } from './worker'
import { Mailer } from './mailer'
import type {
  BuildWorkersParams,
  CreateWorkerConnectionOpts,
  LoadWorkersOpts,
  WorkerBackendOpts,
  WorkerTypes,
  WorkersMode,
} from './types'
import { BullmqBackend } from './backend'

const debug = createDebug('@hyperstackjs/hyperworker')
const load = (patt: string) =>
  glob.sync(patt).map((f) => {
    let mod = null
    try {
      mod = require(f)
    } catch (e: any) {
      console.log(`error dynamically loading '${f}'\n${e.toString()}`) // eslint-disable-line no-console
      throw e
    }
    return mod.default || mod
  })

export const createBackend = ({
  hyperworkers,
  redis,
  backendOpts,
}: {
  hyperworkers: HyperWorker[]
  redis: any
  backendOpts?: WorkerBackendOpts
}) => {
  return new BullmqBackend(hyperworkers, redis, backendOpts)
}

export const loadWorkers = ({ workersGlob, mailersGlob }: LoadWorkersOpts) => {
  const workers = L.reduce(load(workersGlob), L.merge) || {}
  const mailers = L.reduce(load(mailersGlob), L.merge) || {}

  if (
    Object.keys(workers).includes('default') ||
    Object.keys(mailers).includes('default')
  ) {
    throw new Error('cannot start worker that is exported as default')
  }
  return { workers, mailers }
}

export const createConnection = async ({
  redisUrl,
  rejectUnauthorized = false,
}: CreateWorkerConnectionOpts) => {
  const Redis = require('ioredis')
  return new Redis(redisUrl, {
    name: 'hyperworkers',
    lazyConnect: true,
    maxRetriesPerRequest: null,
    enableReadyCheck: false,
    tls: redisUrl?.startsWith('rediss://')
      ? {
          rejectUnauthorized,
        }
      : undefined,
  })
}

const initializeWorkers = ({
  logger,
  mailerTemplatesRoot,
  mailerSettings,
}: {
  logger: any
  mailerTemplatesRoot: string
  mailerSettings: any
}) => {
  HyperWorker.clear()
  Mailer.clear()
  HyperWorker.initWorker({ logger })
  Mailer.initMailer({
    templatesRoot: mailerTemplatesRoot,
    settings: mailerSettings,
  })
}

// recv: beforeWorkers, afterWorkers (like controllers)
const buildWorkers = async (
  {
    isInProcess,
    workersGlob,
    mailersGlob,
    mailerTemplatesRoot,
    mailerSettings,
    logger,
    redisUrl,
    backendOpts,
    truncate = false,
    beforeWorkers = (_p: WorkerTypes) => {},
    afterWorkers = (_p: any) => {},
  }: BuildWorkersParams,
  mode: WorkersMode
) => {
  // worker types rigging

  initializeWorkers({ logger, mailerTemplatesRoot, mailerSettings })

  // set up mail transport and options
  // set up hyperworker logger, report errors with sentry
  beforeWorkers({ HyperWorker, Mailer })

  // load workers from disk
  const { workers, mailers } = loadWorkers({
    workersGlob,
    mailersGlob,
  })

  const numWorkers = Object.keys(workers).length
  const numMailers = Object.keys(mailers).length
  const workerline = `${numWorkers} workers, ${numMailers} mailers`

  const { magenta, bold } = logger.colors
  const pref = bold(magenta('workers:'))

  if (isInProcess) {
    debug('running in-process')
    logger.info(`${pref} [in-process, duplex] ${workerline}`)
    HyperWorker.setMode({ inprocess: isInProcess })
    return {
      background: 'inprocess',
      backend: null,
      connection: null,
      mode: 'duplex' as WorkersMode,
      mailerTemplatesRoot,
      numMailers,
      numWorkers,
      workers,
      mailers,
    }
  }

  const hyperworkers = [
    ...Object.values(workers),
    ...Object.values(mailers),
  ] as HyperWorker[]
  // set up a redis connection
  if (!redisUrl) {
    throw new Error('cannot find a Redis URL')
  }
  debug('creating a Redis connection')
  const redis = await createConnection({ redisUrl })
  await redis.connect()
  debug('connected to Redis')

  logger.info(`${pref} [queue, ${mode}] ${workerline}`)
  if (truncate) {
    debug('truncating Redis content')
    logger.warn(`${pref} truncating Redis contents`)
    await redis.flushdb()
  }

  // build the backend and finish
  const b = createBackend({ hyperworkers, redis, backendOpts })
  debug('created backend')

  // share the backend for message passing, control, or in Redis
  // case, temporary store for workers: backend.connection
  // this is a bit of an oversharing, but very pragmatic
  HyperWorker.backend = b

  afterWorkers({ backend: b })
  if (mode === 'consumer' || mode === 'duplex') {
    debug('processing jobs: consumer mode')
    b.processJobs()
  }

  // afterWorkers({ backend })
  return {
    background: 'queue',
    backend: b,
    connection: redis,
    mailerTemplatesRoot,
    mode,
    numMailers,
    numWorkers,
    workers,
    mailers,
  }
}

export { buildWorkers }
