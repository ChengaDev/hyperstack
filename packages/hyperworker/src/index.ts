import { HyperWorker, producerOnly, queueAs } from './worker'
import { Mailer } from './mailer'
import { buildWorkers } from './hyperworker'
export * from './types'
export * from './ops'

export { HyperWorker, queueAs, Mailer, buildWorkers, producerOnly }
