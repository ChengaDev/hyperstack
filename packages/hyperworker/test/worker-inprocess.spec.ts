import path from 'path'
import { buildWorkers } from '../src'
import type { MailerSettings } from '../src/types'
import { WelcomeMailer } from './fixtures/mailers/welcome'
import { logger } from './utils'
import 'email-templates'

const sleep = (t: number) =>
  new Promise((resolve, _reject) => setTimeout(() => resolve(true), t))
const tieWorkers = async ({
  reportError,
  mailerSettings = undefined,
  beforeWorkers = null,
}: {
  reportError: any
  mailerSettings?: MailerSettings
  beforeWorkers: any
}) => {
  const workersGlob = path.join(__dirname, 'backends', 'fixtures', '*.ts')
  const mailersGlob = path.join(__dirname, 'fixtures', 'mailers', '*.ts')
  const mailerTemplatesRoot = path.join(__dirname, 'fixtures', 'mailers')
  const bw =
    beforeWorkers ||
    (({
      HyperWorker: SomeHW,
      Mailer: SomeMLR,
    }: {
      HyperWorker: any
      Mailer: any
    }) => {
      SomeMLR.setTest()
      SomeHW.setTest(reportError)
    })
  return buildWorkers(
    {
      isInProcess: true,
      workersGlob,
      mailersGlob,
      mailerTemplatesRoot,
      mailerSettings,
      logger,
      beforeWorkers: bw,
    },
    'duplex'
  )
}

describe('worker: inprocess', () => {
  it('should build workers', async () => {
    WelcomeMailer.clearDeliveries()
    const reportError = jest.fn()
    await tieWorkers({ reportError } as any)
    WelcomeMailer.sendWelcome({
      username: 'joe@example.com',
      firstName: 'Smithy',
    }).deliverLater()

    // give template file reading some time, done instead of await'ing the delivery, because in real life controllers we won't be awaiting that.
    // we also lazy load email sending so...
    await sleep(1000)
    expect(WelcomeMailer.deliveries().length).toEqual(1)
    const lastMail = WelcomeMailer.deliveries()[0]

    expect(lastMail.envelope).toMatchSnapshot()
    expect(reportError).toHaveBeenCalledTimes(0)
  })
  it('should not send if send=false', async () => {
    WelcomeMailer.clearDeliveries()
    const reportError = jest.fn()
    await tieWorkers({
      reportError,
      beforeWorkers: () => {}, // avoids setting up test mode
      mailerSettings: { preview: false, send: false },
    })
    WelcomeMailer.sendWelcome({
      username: 'joe@example.com',
      firstName: 'Smithy',
    }).deliverLater()

    // give template file reading some time, done instead of await'ing the delivery, because in real life controllers we won't be awaiting that.
    await sleep(1000)

    expect(WelcomeMailer.deliveries().length).toEqual(0)
    expect(reportError).toHaveBeenCalledTimes(0)
  })
  it('should throw error from mailer direct, when inprocess', async () => {
    WelcomeMailer.clearDeliveries()
    const reportError = jest.fn()
    await tieWorkers({
      reportError,
    } as any)
    WelcomeMailer.badlySendEmail({
      username: 'joe@example.com',
      firstName: 'Smithy',
    }).deliverLater()

    await sleep(300)

    expect(WelcomeMailer.deliveries().length).toEqual(0)
    expect(reportError).toHaveBeenCalledTimes(1)
  })
})
