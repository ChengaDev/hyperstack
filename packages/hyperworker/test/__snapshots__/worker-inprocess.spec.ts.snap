// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker: inprocess should build workers 1`] = `
{
  "from": "no-reply@elle-postage.example.com",
  "to": [
    "joe@example.com",
  ],
}
`;
